datafusion/physical-expr/src/schema_rewriter.rs

```diff
@@ -0,0 +1,318 @@
+                }
+            };
+        let physical_field = self.physical_file_schema.field(physical_column_index);
+
+        let column = match (
+            column.index() == physical_column_index,
+            logical_field.data_type() == physical_field.data_type(),
+        ) {
+            // If the column index matches and the data types match, we can use the column as is
+            (true, true) => return Ok(Transformed::no(expr)),
+            // If the indexes or data types do not match, we need to create a new column expression
+            (true, _) => column.clone(),
+            (false, _) => {
+                Column::new_with_schema(logical_field.name(), self.physical_file_schema)?
+            }
+        };
+
+        if logical_field.data_type() == physical_field.data_type() {
+            // If the data types match, we can use the column as is
+            return Ok(Transformed::yes(Arc::new(column)));
+        }
+
+        // We need to cast the column to the logical data type
+        // TODO: add optimization to move the cast from the column to literal expressions in the case of `col = 123`
+        // since that's much cheaper to evalaute.
+        // See https://github.com/apache/datafusion/issues/15780#issuecomment-2824716928
+        if !can_cast_types(physical_field.data_type(), logical_field.data_type()) {
+            return exec_err!(
+                "Cannot cast column '{}' from '{}' (physical data type) to '{}' (logical data type)",
+                column.name(),
+                physical_field.data_type(),
+                logical_field.data_type()
+            );
+        }
+
+        let cast_expr = Arc::new(CastExpr::new(
+            Arc::new(column),
+            logical_field.data_type().clone(),
+            None,
+        ));
+
+        Ok(Transformed::yes(cast_expr))
+    }
+
+    fn get_partition_value(&self, column_name: &str) -> Option<ScalarValue> {
+        self.partition_fields
+            .iter()
+            .zip(self.partition_values.iter())
+            .find(|(field, _)| field.name() == column_name)
+            .map(|(_, value)| value.clone())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use arrow::datatypes::{DataType, Field, Schema};
+    use datafusion_common::ScalarValue;
+    use std::sync::Arc;
+
+    fn create_test_schema() -> (Schema, Schema) {
+        let physical_schema = Schema::new(vec![
+            Field::new("a", DataType::Int32, false),
+            Field::new("b", DataType::Utf8, true),
+        ]);
+
+        let logical_schema = Schema::new(vec![
+            Field::new("a", DataType::Int64, false), // Different type
+            Field::new("b", DataType::Utf8, true),
+            Field::new("c", DataType::Float64, true), // Missing from physical
+        ]);
+
+        (physical_schema, logical_schema)
+    }
+
+    #[test]
+    fn test_rewrite_column_with_type_cast() -> Result<()> {
```

Comment:
I recommend adding a test for  a more sophisticated expression (like `(a + 5.0::int64) || (c > 0.0::float)`) to ensure that the recursive rewrite case is covered